ngraph/core/reference/src/runtime/reference/experimental_detectron_roi_feature_extractor.cpp

```diff
@@ -0,0 +1,389 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "ngraph/runtime/reference/experimental_detectron_roi_feature_extractor.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <numeric>
+
+#include "ngraph/op/experimental_detectron_roi_feature.hpp"
+#include "ngraph/shape.hpp"
+
+#if defined(__GNUC__) && !defined(__clang__)
+#    if defined(__linux__) && defined(__i386__) && (__GNUC__ == 7 && __GNUC_MINOR__ == 5 && __GNUC_PATCHLEVEL__ == 0)
+#        define NEED_FIX 1
+#    else
+#        define NEED_FIX 0
+#    endif
+#else
+#    define NEED_FIX 0
+#endif
+
+namespace {
+constexpr int64_t input_rois_port = 0;
+constexpr int64_t input_features_start_port = 1;
+
+void redistribute_rois(const std::vector<float>& rois, std::vector<int64_t>& level_ids, const int64_t levels_num) {
+    const float canonical_scale = 224.0f;
+    const int64_t canonical_level = 2;
+    const size_t num_rois = level_ids.size();
+
+    for (size_t i = 0; i < num_rois; ++i) {
+        const float x0 = rois[4 * i + 0];
+        const float y0 = rois[4 * i + 1];
+        const float x1 = rois[4 * i + 2];
+        const float y1 = rois[4 * i + 3];
+
+        int64_t target_level = levels_num;
+        float area = (x1 - x0) * (y1 - y0);
+        if (area > 0) {
+            area = std::sqrt(area) / canonical_scale;
+            area = std::log2(area + 1e-6f);
+            target_level = static_cast<int64_t>(std::floor(area + canonical_level));
+            target_level = std::max(static_cast<int64_t>(0), std::min(levels_num - 1, target_level));
+        }
+
+        level_ids[i] = target_level;
+    }
+}
+
+void reord(const std::vector<float>& src_data,
+           const std::vector<int64_t>& ranks,
+           const int64_t step,
+           float* dst_data,
+           std::vector<int64_t>& dst_mapping) {
+    int64_t n = static_cast<int64_t>(ranks.size());
+
+    std::iota(dst_mapping.begin(), dst_mapping.end(), 0);
```

Comment:
Should we before resize `dst_mapping` to `n`?
It allows to avoid assert in the line 67

Reply:
No, we shouldn't. The function `reord()` is called with such arguments that the condition of the assert is always true.

Reply:
I have deleted this redundant `assert`.